sparse_ho/criterion.py

```diff
@@ -20,178 +20,156 @@ class HeldOutMSE():
-                self.X_test.shape[0])
-        else:
-            self.val_test = None
-
-    def compute_rmse(self, mask, dense, beta_star):
-        if beta_star is not None:
-            diff_beta = beta_star.copy()
-            diff_beta[mask] -= dense
-            self.rmse = norm(diff_beta)
-        else:
-            self.rmse = None
+    def value_outer_crit(self, X, y, mask, dense):
+        """Compute the MSE on the validation set.
+        """
+        if X is None or y is None:
+            return None
+        return norm(y - X[:, mask] @ dense) ** 2 / len(y)
 
-    def get_val(self, log_alpha, tol=1e-3):
+    def get_val(self, model, X, y, log_alpha, tol=1e-3):
         # TODO add warm start
         mask, dense, _ = get_beta_jac_iterdiff(
-            self.model.X, self.model.y, log_alpha, self.model, tol=tol, compute_jac=False)
-        self.value_test(mask, dense)
-        return self.value(mask, dense)
+            X[self.idx_train], y[self.idx_train], log_alpha, model, tol=tol, compute_jac=False)
+        self.get_mse_test(mask, dense)
+        return self.value_outer_crit(mask, dense)
 
     def get_val_grad(
-            self, log_alpha, get_beta_jac_v, max_iter=10000, tol=1e-5,
+            self, model, X, y, log_alpha, get_beta_jac_v, max_iter=10000, tol=1e-5,
             compute_jac=True, beta_star=None):
+
+        def get_v(mask, dense):
+            return 2 * (
+                X[np.ix_(self.idx_val, mask)].T @ (
+                    X[np.ix_(self.idx_val, mask)] @ dense - y[self.idx_val])) / len(self.idx_val)
+
         mask, dense, grad, quantity_to_warm_start = get_beta_jac_v(
-            self.model.X, self.model.y, log_alpha, self.model, self.get_v,
-            mask0=self.mask0, dense0=self.dense0,
+            X[self.idx_train, :], y[self.idx_train], log_alpha, model,
+            get_v, mask0=self.mask0, dense0=self.dense0,
             quantity_to_warm_start=self.quantity_to_warm_start,
             max_iter=max_iter, tol=tol, compute_jac=compute_jac, full_jac_v=True)
         self.mask0 = mask
         self.dense0 = dense
         self.quantity_to_warm_start = quantity_to_warm_start
-        mask, dense = self.model.get_primal(mask, dense)
-        val = self.value(mask, dense)
-        self.value_test(mask, dense)
-        self.compute_rmse(mask, dense, beta_star)
-        if self.convexify:
-            val += self.gamma_convex + np.sum(np.exp(log_alpha) ** 2)
-            grad += 2 * self.gamma_convex * np.exp(log_alpha)
+        mask, dense = model.get_primal(
+            X[self.idx_train, :], y[self.idx_train], mask, dense)
+        val = self.value_outer_crit(
+            X[self.idx_val, :], y[self.idx_val], mask, dense)
+        # TODO put the following in a callback function
+        self.mse_test = self.value_outer_crit(
+            self.X_test, self.y_test, mask, dense)
         return val, grad
 
+    def proj_param(self, model, X, y, log_alpha):
+        return model.proj_param(
```

Comment:
what does proj_param do?

Reply:
it ensure that the parameters ```log_alpha``` does not take value absurdly large, typically we bound it between ```log_alpha_max``` and ```log_alpha_max - 8```

Reply:
Shouldn't it be an attribute of model ?

Reply:
> Shouldn't it be an attribute of model ?

For simple models, yes. For more complex criterions, like the cross validation, or the multiclass it is convenient to handle the projection in the criterion and not in the model.